### PR TITLE
Nano : fix and enhancement of InferenceOptimizer

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

Fix https://github.com/intel-analytics/BigDL/issues/5609, https://github.com/intel-analytics/BigDL/issues/5610, https://github.com/intel-analytics/BigDL/issues/5589

### 1. Why the change?

Fix some bugs and provide some inhancements for InferenceOptimizer

### 2. User API changes

No change.

### 3. Summary of the change 

- [x] fix accuracy KeyError when no metric and validation in optimize
- [x] add unittest for no metric
- [x] add model.eval
- [x] automatically print accelaration results at the end of optimize
- [x] dynamic control of latency_sample_num
- [x] add unittest for get_best_model before optimize
- [x] fix accuracy calculating

### 4. How to test?
- [x] Unit test
- [x] Jenkins
http://10.112.231.51:18889/job/ZOO-PR-NanoTests/384/

### CODE REVIEW

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 1]
```diff
@@ -160,6 +165,8 @@ def optimize(self, model: nn.Module,
 
         result_map: Dict[str, Dict] = {}
 
+        model.eval()  # change model to eval state
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:168]: Can you first record model's status, and restore it after we are done?

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 2]
```diff
@@ -420,15 +440,22 @@ def _available_acceleration_combination():
...
     '''
+    start_time = time.perf_counter()
     time_list = []
-    for _ in range(iterrun):
-        st = time.time()
+    for i in range(iterrun):
+        st = time.perf_counter()
         func(*args)
-        time_list.append(time.time() - st)
+        end = time.perf_counter()
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:450]: This perf counter might be a good bkm for future time measurement

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py [hunk 3]
```diff
@@ -69,25 +69,54 @@ class TestInferencePipeline(TestCase):
...
+        acc_model(next(iter(self.train_loader))[0])
+
+    def test_pipeline_without_metric(self):
+        inference_opt = InferenceOptimizer()
+        inference_opt.optimize(model=self.model,
+                               training_data=self.train_loader,
+                               cpu_num=1)
+
+        acc_model, option = inference_opt.get_best_model()
+        print(option)
```
Comment (TheaperDeng) [inline python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:106]: no need to print

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py [hunk 4]
```diff
@@ -69,25 +69,54 @@ class TestInferencePipeline(TestCase):
...
+        acc_model, option = inference_opt.get_best_model(accelerator="onnxruntime")
+        assert option == "" or "onnxruntime" in option
+        acc_model, option = inference_opt.get_best_model(precision="int8")
+        assert option == "" or "inc" in option
```
Comment (TheaperDeng) [inline python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:110]: `assert option == "" or "inc" in option or "pot" in option`?

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py [hunk 5]
```diff
@@ -69,25 +69,54 @@ class TestInferencePipeline(TestCase):
...
+        assert error_msg == "If you want to specify accuracy_criterion, you need "\
+                            "to set metric and validation_data when call 'optimize'."
+
 
 if __name__ == "__main__":
     test = TestInferencePipeline()
-    test.test_pipeline()
+    test.test_get_model_without_optimize()
+    test.test_pipeline_with_metric()
+    test.test_pipeline_without_metric()
```
Comment (TheaperDeng) [inline python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:122]: no need to put this main program here, pytest will capture all the legal unittest automatically.

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py (hunk 3 above)
Comment (rnwang04) [reply python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:106]: done

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py (hunk 5 above)
Comment (rnwang04) [reply python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:122]: done

File: python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py (hunk 4 above)
Comment (rnwang04) [reply python/nano/test/pytorch/tests/test_inference_pipeline_ipex.py:110]: Good catch! done

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py (hunk 1 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:169]: Yes. After update, now I record model's status by `model.training` at the beginning of `optimize`, and restore it at the end of `optimize`.
So if a model is in training status, after `optimize`, it is still in training status, but all accelarated models will be in eval status always.

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 6]
```diff
@@ -228,17 +236,31 @@ def func_test(model, input_sample):
...
+                else:
+                    result_map[method]["accuracy"] = None
                 result_map[method]["model"] = acce_model
 
             else:
                 pass
 
+        # restore model's state
+        if training_state:
+            model.train()
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:252]: should we consider if `self.optimized_model_dict["original"]["model"]` be set to training model? Since we directedly set model without deepcopy https://github.com/intel-analytics/BigDL/blob/b455568a73ca92683bd402c2a617bd15a40ab775/python/nano/src/bigdl/nano/pytorch/inference/optimizer.py#L183 that's not good for users.

Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:252]: Good catch! Add deepcopy. Now only the original model will be set to training mode(if it is in training mode before optimize) and all accelerated model(include "original") will be set in eval mode.

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 7]
```diff
@@ -179,19 +180,19 @@ def optimize(self, model: nn.Module,
                     input_sample = tuple(next(iter(training_data))[:-1])
                     try:
                         if accelerator is None and use_ipex is False:
-                            acce_model = model
+                            acce_model = deepcopy(model)
```
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:183]: Here I use deepcopy to make `self.optimized_model_dict["original"]["model"]` always in eval mode no matter whether original model in training mode.
If remove deepcopy, when user pass a model in training mode,  `self.optimized_model_dict["original"]["model"]` will also be training mode. So It cannot be guaranteed that the models saved by optimize are all eval model. Is this behavior reasonable?

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:183]: Will this be an issue if you are optimizing, say, BERT-Large? Will making a deep copy be a problem?

It is probably reasonable that the "original" model is the same as the original one?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:183]: Users may get accelerated model from `optimizer.get_best_model(...)`, but if all accelerated method is not valid (either failed or does not meet user's accuracy requirement), we will let users get the original model.

Users may familiar this usage:
```python
acc_model = optimizer.get_best_model()
acc_model(x)
```

So, that's why we want to make the returned model in eval mode.

Well, it's just a minor point and it may does cause some mem issue. maybe we will just note our user the original model is the same reference as the model before optimization (i.e. they have the same training/eval mode)

[the remark this answers is not in the document]
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:183]: > Will this be an issue if you are optimizing, say, BERT-Large? Will making a deep copy be a problem in this case?
> 
> It is probably reasonable that the "original" model is the same as the original one?

Yeah, I agree. I will remove this deepcopy, and we will maybe note our user the original model is the same reference as the model before optimization as junwei comment.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:183]: > Users may get accelerated model from `optimizer.get_best_model(...)`, but if all accelerated method is not valid (either failed or does not meet user's accuracy requirement), we will let users get the original model.
> 
> Users may familiar this usage:
> 
> ```python
> acc_model = optimizer.get_best_model()
> acc_model(x)
> ```
> 
> So, that's why we want to make the returned model in eval mode.
> 
> Well, it's just a minor point and it may does cause some mem issue. maybe we will just note our user the original model is the same reference as the model before optimization (i.e. they have the same training/eval mode)

Maybe a simpler semantics is to let the user know that `InferenceOptimizer` will always set all models (including the original one) to `eval`?

[the remark this answers is not in the document]
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py]: So we don't restore model's status at the end of `optimize`, and just remind user that all models will be set to `eval`？

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py]: Yes, I think it's better than deepcopy; to the user, `optimize` will always set `eval`, which is easier to understand.

[the remark this answers is not in the document]
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py]: Done.
